[Service Bus] Add more tests for management API

ramya0820 opened the pull request:
For more context refer to #4509


From https://github.com/Azure/azure-sdk-for-js/issues/4509#issuecomment-549657022 notes inline and bold:

_From @ramya-rao-a: Update: Comments in italics represent conclusion after offline discussions_

- atomManagement.spec.ts
   - Updating non existing entity should fail 
      **added**
   - Listing when no entities exist should return empty array in the response 
      **added**
   - Passing invalid options returns expected error in places where we have added checks and are returning type errors
      **these result in compilation errors and can't be included as discussed earlier. Although, having JS specific tests are useful and should be something we might want to pick up for across SDKs.**
      _@ramya-rao-a: Compilation errors can be worked around by using `as any`. This ask was about test coverage for places where we, the library throws error while validating user input_
   - Listing with skip & top options
      **added**
   - existing create & update tests do not check if the options that are passed are actually getting set or not.
      **as discussed, all properties tests exist for create - added similar coverage for update**
   - update existing test cases for creating an existing entity or deleting non existing entity to check error message
      **added, as discussed the focus is more on parsing structure and avoiding taking a hard dependency on error message strings**
      _@ramya-rao-a: It was concluded that we skip error message check when error is coming from service. We keep error message checks for errors that we explicitly throw_

- atomXml.spec.ts
   - pass valid xml content to `deserializeAtomXmlResponse`, but with no `feed` or `entry` elements and test that expected error is thrown. 
     **added**
   - pass response with unrecognized status code to `deserializeAtomXmlResponse`, and test that expected error is thrown
     **added**

- serializer helper method tests
   - pass options (not all) out of order to `buildQueueOptions` and test that the output should consist all fields and is in expected order. Same for topic, subscription and rule. 
     **this test seems better written against core-http since it's a general one about if for given JSON, the XML tags are in expected output**
    _@ramya-rao-a: This is about getting test coverage for the code that we have that manipulates user given input data into a particular order and differently cased property names. Therefore, this is still a valid case for getting more test coverage_
   - pass invalid content for authorization rules `buildQueueOptions` and test that expected error is thrown. Same for topic and subscription
     **this would be testing the service and seems extraneous**
    _@ramya-rao-a: This is not about testing the service. This is about testing the validation checks we have in place for the auth rules provided by the user._
   - pass valid content (with all properties) to `buildQueue` and test that output is in expected shape. Same for topic, subscription and rule
     **this coverage already exists with all properties tests**
   - pass invalid content for sql parameters in sql filter to the rule serializer's serialize() and test that expected error is thrown. Same for sql action.
     **this would be testing the service and seems extraneous**
    _@ramya-rao-a: This is not about testing the service. This is about testing the validation checks we have in the serialize() method that checks user provided data_

chradek commented on sdk/servicebus/service-bus/test/atomXml.spec.ts:
```diff
@@ -72,6 +76,59 @@ describe("atomSerializationPolicy #RunInBrowser", function() {
+describe("deserializeAtomXmlResponse #RunInBrowser", function() {
+  it("should throw an error if receiving a valid XML but invalid Atom XML", async function() {
+    try {
+      await serviceBusAtomManagementClient.deleteQueue("alwaysBeExistingQueue");
```
Actually, for this test you probably don't even need to make a real request. You can just create a `_response` object that has the fields you need set (status, parsedBody, request if needed).

ramya0820 replied:
That's what I had first tried but kept getting a 'cannot read property clone() on undefined' error.
Still not sure where that is coming from..

ramya0820 replied:
Looked into this a little more, looks like it needed the headers/request to be constructed properly - updated to use mocked functionality

chradek commented on sdk/servicebus/service-bus/test/atomXml.spec.ts:
```diff
@@ -72,6 +76,59 @@ describe("atomSerializationPolicy #RunInBrowser", function() {
+    }
+    const response = await serviceBusAtomManagementClient.createQueue("alwaysBeExistingQueue");
+
+    response._response.status = 666;
```
Same here, I don't think you actually have to make requests to the service if you create the `_response` yourself. There aren't really 'live' tests anyway.

chradek commented on sdk/servicebus/service-bus/src/util/utils.ts:
```diff
@@ -302,6 +302,13 @@ function buildAuthorizationRule(value: any): AuthorizationRule {
+
+  if (
+    authorizationRule.rights.accessRights &&
+    !Array.isArray(authorizationRule.rights.accessRights)
```
I just noticed line 299
`accessRights: value["Rights"]["AccessRights"]`
This could cause an exception if `value["Rights"] is undefined. Since you're touching this function already can you fix that as well?

ramya0820 replied:
updated

chradek replied:
Thank you!

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -228,6 +236,118 @@ const alwaysBeDeletedRule = "alwaysbedeletedrule";
 
+      it(`Update on non-existent ${entityType} entity throws an error`, async () => {
+        let response;
+        switch (entityType) {
```
Have you thought about instead of passing in an array of `EntityType` and performing these case statements in each `it` scope, passing in an array of objects that have the additional fields that your switch statements are using?

For example, something like: 
```js
[{
  entityType: EntityType.QUEUE,
  entityNames: [alwaysBeExistingQueue]
}, {
  entityType: EntityType.SUBSCRIPTION,
  entityNames: [alwaysBeExistingSubscription, alwaysBeExistingTopic]
}, /* etc */
].forEach((testDetails) => {})
```

It just seems like if you're already going through the trouble of reusing an `it` because the testing is similar for each resource, you're losing out on the real benefit since you then check the EntityType in a switch statement and have all your logic in each case. With the way it is now I don't see how this is better than having separate `it` tests that are focused for each resource. (I'd argue it's worse now since resource-specific logic for every resource is in a single `it`, which makes it harder to understand what's going on)

ramya0820 replied:
Agree with simplifying the switch case within the `it`s - updated

ramya-rao-a commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -162,6 +162,14 @@ const alwaysBeDeletedRule = "alwaysbedeletedrule";
+          response.message.startsWith("The messaging entity") ||
+            response.message.startsWith("Entity") ||
+            response.message.startsWith("SubCode") ||
+            response.message.startsWith("No service is hosted at the specified address."),
```
This tells us that creating a non existing entity results in 4 different types of error messages. Are the 4 different types of messages based on the entity type or any other scenario?

@bterlson, would you consider the client library to step in here and provide a consistent error message or we throw whatever the service decided to throw?

bterlson replied:
I don't want to second guess the service here, personally, but could be convinced otherwise

ramya0820 replied:
The service sends us error information that could grammatically / structurally be constructed in any of those 4 known ways I would say.. (Shortened the last message string to convey this intent.)

We had specifically discussed about trying best to not take a hard dependency on error message sentences and that intent of testing error messages were mainly for verifying the error structure parsing - hence the tests were updated accordingly.

ramya-rao-a commented on sdk/servicebus/service-bus/test/atomXml.spec.ts:
```diff
@@ -72,6 +76,59 @@ describe("atomSerializationPolicy #RunInBrowser", function() {
+      assert.deepEqual(true, false, "Error must be thrown");
+    } catch (err) {
+      assert.deepEqual(
+        err.message.startsWith("Service Error"),
```
Since we the client control this error message, can we not check for the whole message instead of just what it starts with? Same for the check for error code

ramya0820 replied:
updated

ramya-rao-a commented:
@ramya0820 Does this PR add all the tests referred to in https://github.com/Azure/azure-sdk-for-js/issues/4509#issuecomment-549657022? If not, please update the PR description with a copy of the items from the above comment that are addressed in this PR

ramya0820 commented on sdk/servicebus/service-bus/src/serviceBusAtomManagementClient.ts:
```diff
@@ -352,6 +353,12 @@ export class ServiceBusAtomManagementClient extends ServiceClient {
       `Performing management operation - updateQueue() for "${queueName}" with options: ${queueOptions}`
     );
 
+    if (!isJSONLikeObject(queueOptions) || queueOptions == undefined) {
```
updated

ramya0820 commented on sdk/servicebus/service-bus/test/atomXml.spec.ts:
```diff
@@ -72,6 +76,59 @@ describe("atomSerializationPolicy #RunInBrowser", function() {
+
+    delete response._response["parsedBody"]["entry"];
+    delete response._response["parsedBody"]["feed"];
+    response._response["parsedBody"] = { notAnEntry: "" };
```
updated - I was testing few things, this got left out

ramya0820 commented on sdk/servicebus/service-bus/test/atomXml.spec.ts:
```diff
@@ -72,6 +76,59 @@ describe("atomSerializationPolicy #RunInBrowser", function() {
+    try {
+      await serviceBusAtomManagementClient.deleteQueue("alwaysBeExistingQueue");
+    } catch (err) {
+      console.log("Ignoring clean up step");
```
updated, using the logger so that these can be enabled if needed for tests

ramya0820 commented:
> @ramya0820 Does this PR add all the tests referred to in [#4509 (comment)](https://github.com/Azure/azure-sdk-for-js/issues/4509#issuecomment-549657022)? If not, please update the PR description with a copy of the items from the above comment that are addressed in this PR

Updated PR description with response to the comment. 
Discussed with @AlexGhiondea earlier on this - about list entities, initially it was assumed that we wouldn't be able to reuse same namespace, but added in cleanup tests that would delete all queues and topics before the tests are run instead.
About other tests, added in notes on why they weren't included in current PR. @ramya-rao-a

chradek commented on sdk/servicebus/service-bus/src/serviceBusAtomManagementClient.ts:
```diff
@@ -748,7 +775,27 @@ export class ServiceBusAtomManagementClient extends ServiceClient {
     const webResource: WebResource = new WebResource(this.getUrl(name), "GET");
 
-    return executeAtomXmlOperation(this, webResource, serializer);
+    return new Promise<HttpOperationResponse>(async (resolve, reject) => {
```
There shouldn't be a need to wrap this in a new Promise since `getResource` is already an async function. You should be able to do refactor like this:

```ts
// don't need try/catch around this because we want the error to be thrown anyway
const response = await executeAtomXmlOperation(this, webResource, serializer);
if (
  response.parsedBody == undefined ||
  (Array.isArray(response.parsedBody) && response.parsedBody.length == 0)
) {
  const err = new RestError(
    `The messaging entity "${name}" being requested cannot be found.`,
    "404",
    404,
    stripRequest(webResource),
    stripResponse(response)
  );
  throw err;
}
return response;
```

ramya0820 replied:
Updated

ramya0820 commented:
About the tests on serializers and testing the XML content being serialized to, 
- one aspect as added in description is to test that the XML tags are generated in order of input JSON, and would be a generic test better added in core-http. (As it mainly tests the XML serialization helper tool).
- the other aspect which I think we maybe referring to is to safeguard against changes to the serializers that may change the order of XML tags by having in unit tests that ensure the ranking/order of properties in output XML are same and as defined exactly?
    - This doesn't seem right/useful as there is no exact order to use and the order can be changed in few places if needed. Some changes in order result in unpredictable behavior from service and are bugs. So far we have addressed this by working backwards from the functionality we want to see as working (since no clear contract/documentation exists for the service side impl.) 
Thus, the exact ordering of fields is not known/defined and is a service side implementation detail that I think shouldn't be explicitly maintained in the SDK as such. Thoughts?

cc @ramya-rao-a @AlexGhiondea @bterlson @chradek

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
+  }
+].forEach((testCase) => {
+  describe(`Atom management - Clean namespace for "${testCase.entityType}"`, function(): void {
+    it(`Deletes all ${testCase.entityType} entities successfully`, async () => {
```
Is there any reason we need to delete _every_ entity in the namespace? It looks like the intent is to ensure we can delete an entity. What if instead we create an entity, wait for it to exist, then delete it and verify it no longer exists?

Right now it looks like this test could have a lot of false positives. If `listEntities` returns an empty list, we'll never attempt to delete something. If `listEntities` or `deleteEntity` fails, we catch the error and don't assert anything. It doesn't look like it's possible for this test to fail.

ramya0820 replied:
Yes, this was the "clean namespace" objective we had discussed in meeting.
I intended for this to run before the management API tests run.
It's a hacky workaround to be able to test the feature/API, otherwise it was not being made possible to be able to test the list() APIs with current setup of test infra.

As discussed, we will be removing the list() tests out of current PR, until we have decided on using an empty namespace or some other approach. https://github.com/Azure/azure-sdk-for-js/issues/6197

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
-      it(`Lists available ${entityType} entities successfully`, async () => {
-        let response;
+      try {
+        await createEntity(
```
I'm not sure why you're creating an entity here since the test is checking that `listEntities` returns a result if no entities exist, and nothing is asserted on `createEntity` anyway.

ramya0820 replied:
Idea was that "alwaysBeExisting" entities are always existing for tests that rely on their existence. 

These tests will be removed though as discussed in yesterdays meeting.

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
+      should.equal(
+        response.length,
+        testCase.entityType == EntityType.RULE ? 1 : 0,
+        "Result must be an empty array except for rules"
```
Just curious, why is rules non-empty? Does it have a default that's always returned?

ramya0820 replied:
Yeah, one a default one gets created with name "$Default"

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
-              alwaysBeExistingSubscription
-            );
-            break;
+      try {
```
I think it could be useful to remove the try/catch on both of these createEntity calls. If either fails, it would help us know in CI that this was the root cause for the test failing.

Also, you can have a false positive currently if the 2nd createEntity fails, since then there'd only be one result to return.

ramya0820 replied:
Updated, also see - https://github.com/Azure/azure-sdk-for-js/pull/6112#issuecomment-555716170

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
-        }
-      });
+      try {
+        await deleteEntity(
```
It's starting to look like a lot of these tests depend on all entities being deleted before you start running them. If this is the case, I'd recommend using an `afterEach` or `beforeEach` block to delete all your entities. This ensures that a failure in one test doesn't leak state into another test (for example, if your `listEntities` call threw an error, then `deleteEntity` would never be called.

ramya0820 replied:
Updated

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
-              alwaysBeExistingSubscription
-            );
-            break;
+      try {
```
Same comment about the try/catch as above. Would be helpful to remove to reduce false positives/make failures in CI easier to troubleshoot.

ramya0820 replied:
Updated, also see https://github.com/Azure/azure-sdk-for-js/pull/6112#issuecomment-555716170

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
+    it(`Lists available ${testCase.entityType} entities successfully`, async () => {
+      let response;
+
+      response = await listEntities(
```
I think we should create some entities first before listing to be sure there are entities to list. If I ran this test in isolation (it.only) it should work.

Then we can also test that the array length is > 0. (We already have a test that ensures we can get an empty list, would be good to verify we can get a non-empty list which I think is what this test is for).

ramya0820 replied:
Updated

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
+    it(`Updates an existent ${testCase.entityType} entity successfully`, async () => {
+      let response: any;
+
+      response = await updateEntity(
```
How do we know the entity is already existing?

ramya0820 replied:
Updated, all of these rely on the "alwaysBeExisting.."

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
+    it(`Gets an existent ${testCase.entityType} entity successfully`, async () => {
+      let response;
+
+      response = await getEntity(
```
Same here, how do we know the entity already exists?

ramya0820 replied:
https://github.com/Azure/azure-sdk-for-js/pull/6112#issuecomment-555716170

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
+    it(`Deletes an existent ${testCase.entityType} entity successfully`, async () => {
+      let response;
+
+      try {
```
Same comment about removing try/catch.

ramya0820 replied:
If a test fails on CI, then manual scripts would need to be run - or the "clean namespace" mechanism we have should be updated to delete all the entities.
For such cases (specifically, simple delete and create tests), had added in the ignorable clean up steps. 
However, updated it now to use isolation as the CI runs are generally expected to be clean

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -47,364 +48,444 @@ const alwaysBeDeletedSubscription = "alwaysbedeletedsubscription";
+      } catch (err) {
+        response = err;
+      }
+      try {
```
Why do we create an entity here?

ramya0820 replied:
Mocha doesn't allow for specific before/after to be added, thus ones that are test specific are included within the test

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -560,8 +641,12 @@ const alwaysBeDeletedRule = "alwaysbedeletedrule";
 });
 
 // Topic tests
+// Create different topic each testcase as updates to same topic/subscription does not scale on service side and gives us error -
```
Appreciate the comment!

ramya0820 replied:
Thanks and yes! This is actually one other reason why I would refrain from the afterEach beforeEach approach - because the service may not be able to handle.

This was quite a mess to deal with leading to specific test setup such as above.

So, the more I think about it - it may be very hard to achieve the level of test isolation we are looking for. I'll do one pass of test refactoring to see if things can be improved without breaking.

ramya0820 replied:
See https://github.com/Azure/azure-sdk-for-js/pull/6112#issuecomment-555716170

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -765,17 +855,22 @@ const alwaysBeDeletedRule = "alwaysbedeletedrule";
 ].forEach((testCase) => {
   describe(`Subscription creation with differing options #RunInBrowser`, function(): void {
     it(`${testCase.testCaseTitle}`, async () => {
+      try {
```
Same comment earlier about possibly moving entity deletion to a beforeEach or afterEach block.

ramya0820 replied:
Updated

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -765,17 +855,22 @@ const alwaysBeDeletedRule = "alwaysbedeletedrule";
   }
 ].forEach((testCase) => {
   describe(`Subscription creation with differing options #RunInBrowser`, function(): void {
     it(`${testCase.testCaseTitle}`, async () => {
```
This is a very hard to understand description just looking at the code. Can you add any additional text around the title you're injecting? It's not clear what the test is actually supposed to be testing.

ramya0820 replied:
"Subscription creation with differing options" -> 

createSubscription() using different variations to the input parameter "subscriptionOptions"

Would this be more clear?

chradek replied:
I think that's clear enough for me. Yeah, it was just seeing `differing options` that I thought was ambiguous, maybe that's just me.

ramya0820 replied:
Updated

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -935,38 +1016,499 @@ const alwaysBeDeletedRule = "alwaysbedeletedrule";
+      // enableSubscriptionPartitioning: true,
+      // filteringMessagesBeforePublishing: true,
+
+      // maxSizeInMegabytes: 2048, // For partitioned entities, this is 16384
```
Should these lines be removed, or are they something we plan to add later? (If the latter, can you add a comment indicating these should be added in the future with a tracking issue?)

ramya0820 replied:
Added

chradek replied:
Sorry, I don't see the comment pointing to the tracking issue.

ramya0820 commented on sdk/servicebus/service-bus/src/serviceBusAtomManagementClient.ts:
```diff
@@ -748,7 +775,21 @@ export class ServiceBusAtomManagementClient extends ServiceClient {
+      response.parsedBody == undefined ||
+      (Array.isArray(response.parsedBody) && response.parsedBody.length == 0)
+    ) {
+      const err = new RestError(
```
Yes, since we decided to treat service response of 200 for these to be a bug that we want to cover up for.

ramya0820 commented:
- Tests have been updated to improve the isolation. One more pass of refactoring is pending to see if the 2-3 remaining tests can be further isolated. 
- The parameter validation tests have been added - can review these along with other unit tests and look at live tests refactoring to later.
@chradek

ramya0820 commented:
Another point discussed offline about the test coverage for serializers is that additional ones are added to test the ordering in output XML (based on given JSON input) which is the actual contract with service that we are reverse engineering and implementing - this would also help catch if anything changed with JS/TS treatment of objects or xml parsing utility used in general. cc: @chradek

ramya0820 commented:
Looks like master got updated while the merge was in progress -.-
Updating it in a bit

ramya0820 commented:
Changes have been updated from master. The issue with using JSON.stringify/parse was fixed on master and resulted in few more changes to `atomXml.spec.ts`.

Other than that, https://github.com/azure/azure-sdk-for-js/issues/6219 has been filed for further improvement on tests and input validation.
https://github.com/azure/azure-sdk-for-js/issues/6276 has been filed to see if further improvements can be made to the tests structure.

cc @chradek

chradek commented on sdk/servicebus/service-bus/test/atomManagement.spec.ts:
```diff
@@ -35,43 +35,32 @@ enum EntityType {
   }
 ].forEach((testCase) => {
-  describe(`Atom management - Basic CRUD on "${testCase.entityType}" entities #RunInBrowser`, function(): void {
+  describe.only(`Atom management - Basic CRUD on "${testCase.entityType}" entities #RunInBrowser`, function(): void {
```
Remove .only please :)

This actually appears in multiple spots. Please remove all `only`s!

ramya0820 replied:
Oops, change got excluded from local :/ - removed them

chradek commented:
/azp run js - servicebus - tests
